osgp/platform/osgp-secret-management/src/main/java/org/opensmartgridplatform/secretmanagement/application/services/SecretManagementService.java

```diff
@@ -0,0 +1,164 @@
+/**
+ * Copyright 2020 Smart Society Services B.V.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package org.opensmartgridplatform.secretmanagement.application.services;
+
+import java.util.Date;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import org.apache.tomcat.util.buf.HexUtils;
+import org.opensmartgridplatform.secretmanagement.application.domain.DbEncryptedSecret;
+import org.opensmartgridplatform.secretmanagement.application.domain.DbEncryptionKeyReference;
+import org.opensmartgridplatform.secretmanagement.application.domain.SecretType;
+import org.opensmartgridplatform.secretmanagement.application.domain.TypedSecret;
+import org.opensmartgridplatform.secretmanagement.application.repository.DbEncryptedSecretRepository;
+import org.opensmartgridplatform.secretmanagement.application.repository.DbEncryptionKeyRepository;
+import org.opensmartgridplatform.shared.security.EncryptedSecret;
+import org.opensmartgridplatform.shared.security.EncryptionDelegate;
+import org.opensmartgridplatform.shared.security.EncryptionProviderType;
+import org.opensmartgridplatform.shared.security.Secret;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.Pageable;
+import org.springframework.stereotype.Service;
+
+@Service
+public class SecretManagementService {
+    private final EncryptionDelegate encryptionDelegate;
+    private final EncryptionProviderType encryptionProviderType;
+    private final DbEncryptedSecretRepository secretRepository;
+    private final DbEncryptionKeyRepository keyRepository;
+
+    public SecretManagementService(
+            @Qualifier("DefaultEncryptionDelegate") final EncryptionDelegate defaultEncryptionDelegate,
+            final EncryptionProviderType encryptionProviderType, final DbEncryptedSecretRepository secretRepository,
+            final DbEncryptionKeyRepository keyRepository) {
+        this.encryptionDelegate = defaultEncryptionDelegate;
+        this.encryptionProviderType = encryptionProviderType;
+        this.secretRepository = secretRepository;
+        this.keyRepository = keyRepository;
+    }
+
+    public void storeSecrets(final String deviceIdentification, final List<TypedSecret> secrets) throws Exception {
+        //@formatter:off
+        secrets.stream()
+                .map(t -> this.validateSecret(deviceIdentification, t))
+                .map(s -> this.createEncrypted(deviceIdentification, s, this.getKey()))
+                .forEach(this.secretRepository::save);
+        //@formatter:on
+    }
+
+    private DbEncryptionKeyReference getKey() {
+        final Date now = new Date();
```

Comment:
Avoid using the old java.util.Date class. ZonedDateTime is a better one.

Reply:
The date is also used in the call to the "find" for the DB, so using java.util.Date is still necessary. I'll mark this comment as resolved.